pkg/driver/oxide/oxide.go

```diff
@@ -0,0 +1,423 @@
+		{&d.Token, flagToken},
+		{&d.Project, flagProject},
+	}
+
+	for _, f := range flags {
+		*f.target = opts.String(f.flag)
+		if *f.target == "" {
+			return requiredOptionError(f.flag)
+		}
+	}
+
+	d.Host = opts.String(flagHost)
+	d.Token = opts.String(flagToken)
+	d.Project = opts.String(flagProject)
+	d.UserData = opts.String(flagUserData)
+	d.ExtraUserData = opts.String(flagExtraUserData)
+
+	d.DiskSize = opts.Int(flagDiskSize)
+	d.DiskImageId = opts.String(flagDiskImageId)
+	d.Memory = opts.Int(flagMemory)
+	d.Ncpus = opts.Int(flagNCpus)
+
+	d.VpcName = opts.String(flagVpc)
+	d.SubnetName = opts.String(flagSubnet)
+
+	d.BaseDriver.SSHUser = opts.String(flagSSHUser)
+	d.SSHKeys = opts.StringSlice(flagSSHPubKeys)
+
+	// base driver flags
+	d.BaseDriver.SSHPort = 22
+	d.SetSwarmConfigFromFlags(opts)
+
+	return
+}
+
+// PreCreateCheck is called to enforce pre-creation steps
+func (d *Driver) PreCreateCheck() error {
+	if len(d.userDataBase64())*3/4 > 32767 {
+		return fmt.Errorf("user data is too large")
+	}
+	return nil
+}
+
+func (d *Driver) Create() (err error) {
+	var ctx = context.Background()
+	var networkInterfaces []oxide.InstanceNetworkInterface
+	var sshKey *oxide.SshKey
+
+	if err = d.SetupClient(); err != nil {
+		return
+	}
+
+	if sshKey, err = d.SshSetup(); err != nil {
+		return
+	}
+
+	sshKeys := []oxide.NameOrId{
+		oxide.NameOrId(sshKey.Id),
+	}
+
+	for _, key := range d.SSHKeys {
+		sshKeys = append(sshKeys, oxide.NameOrId(key))
+	}
+
+	if _, err = d.Client.InstanceCreate(ctx,
+		oxide.InstanceCreateParams{
+			Project: oxide.NameOrId(d.Project),
+			Body: &oxide.InstanceCreate{
+				Description: "built by rancher-machine driver",
+				BootDisk: &oxide.InstanceDiskAttachment{
+					Description: "docker-machine disk",
+					DiskSource: oxide.DiskSource{
+						ImageId: d.DiskImageId,
+						Type:    oxide.DiskSourceTypeImage,
+					},
+					Name: oxide.Name(d.diskName()),
+					Size: oxide.ByteCount(d.DiskSize),
+					Type: oxide.InstanceDiskAttachmentTypeCreate,
+				},
+				Disks:    []oxide.InstanceDiskAttachment{},
+				Hostname: oxide.Hostname(d.MachineName),
+				Memory:   oxide.ByteCount(d.Memory),
+				Name:     oxide.Name(d.MachineName),
+				Ncpus:    oxide.InstanceCpuCount(d.Ncpus),
+				NetworkInterfaces: oxide.InstanceNetworkInterfaceAttachment{
+					Type: oxide.InstanceNetworkInterfaceAttachmentTypeCreate,
+					Params: []oxide.InstanceNetworkInterfaceCreate{
+						{
+							Description: "created by foo in rancher-machine driver",
+							Name:        "huh",
+							SubnetName:  oxide.Name(d.SubnetName),
+							VpcName:     oxide.Name(d.VpcName),
+						},
+					},
+				},
+				UserData:      d.userDataBase64(),
+				SshPublicKeys: sshKeys,
+			},
+		},
+	); err != nil {
+		return
+	}
+
+	if networkInterfaces, err = d.Client.InstanceNetworkInterfaceListAllPages(ctx,
+		oxide.InstanceNetworkInterfaceListParams{
+			Project:  oxide.NameOrId(d.Project),
+			Instance: oxide.NameOrId(d.MachineName),
```

Comment:
Instance names (and all names of resources in general) are intended to be mutable in the future. It's a safer bet to retrieve and use the IDs instead. If possible, it be good to change all `List`, `View`, and `Delete` calls to use IDs.

Reply:
From what I can tell while building this, the naming comes from the driver. They are named uniquely though of a form <clusterName>-<poolName>-<deploymentID>-<instanceID> like test-worker-263a05fb-v44b6

Reply:
i.e. the Rancher management cluster holds the source of truth for the ID